File: 189-rotate-array/rotate-array.cpp
#include <algorithm> // For std::reverse

class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        int n = nums.size();
        k = k % n; // Handle cases where k > n

        // Reverse the entire array
        reverse(nums.begin(), nums.end());

        // Reverse the first k elements
        reverse(nums.begin(), nums.begin() + k);

        // Reverse the remaining elements
        reverse(nums.begin() + k, nums.end());
    }
};

File: 189-rotate-array/rotate-array.java
class Solution {
    public void rotate(int[] nums, int k) {
        int n = nums.length;
        k = k % n;

       rev(nums,0,n-1);
       rev(nums,0,k-1);
       rev(nums,k,n-1);
    }
    public static void rev(int[] nums, int i, int j){
        while(j> i){
            int t = nums[i];
            nums[i] = nums[j];
            nums[j] = t;
            i++;
            j--;
        }
    }
}